bullet/src/SDFFeatures.cc

```diff
@@ -69,23 +75,101 @@ Identity SDFFeatures::ConstructSdfLink(
+    linkInertiaDiag = btVector3(0,0,0);
+  }
   btDefaultMotionState* myMotionState = new btDefaultMotionState(baseTransform);
   btRigidBody::btRigidBodyConstructionInfo rbInfo(mass, myMotionState, groundShape, linkInertiaDiag);
   btRigidBody* body = new btRigidBody(rbInfo);
   body->setWorldTransform(baseTransform);
 
-  // Add it to the internal world:
-  const auto _worldID = this->models.at(_modelID)->world;
-  const auto &world = this->worlds.at(_worldID)->world;
-  world->addRigidBody(body);
-
   // Generate an identity for it
   const auto linkIdentity = this->AddLink({name, body, _modelID});
 
+  for (std::size_t i = 0; i < _sdfLink.CollisionCount(); ++i)
+  {
+    this->ConstructSdfCollision(linkIdentity, *_sdfLink.CollisionByIndex(i));
+  }
+
   return linkIdentity;
 }
 
+Identity SDFFeatures::ConstructSdfCollision(
+      const Identity &_linkID,
+      const ::sdf::Collision &_collision)
+{
+  if (!_collision.Geom())
+  {
+    ignerr << "The geometry element of collision [" << _collision.Name() << "] "
+           << "was a nullptr\n";
+    return this->GenerateInvalidId();
+  }
+
+  const auto &geom = _collision.Geom();
+  btCollisionShape* shape = nullptr;
+
+  if (geom->BoxShape())
+  {
+    const auto box = geom->BoxShape();
+    const auto size = math::eigen3::convert(box->Size());
+    const auto halfExtents = convertVec(size)*0.5;
+    shape = new btBoxShape(halfExtents);
+  }
+  else if (geom->SphereShape())
+  {
+    const auto sphere = geom->SphereShape();
+    const auto radius = sphere->Radius();
+    shape = new btSphereShape(radius);
+  }
+  else if (geom->CylinderShape())
+  {
+    const auto cylinder = geom->CylinderShape();
+    const auto radius = cylinder->Radius();
+    const auto halfLength = cylinder->Length()*0.5;
+    shape = new btCylinderShapeZ(btVector3(radius, 0, halfLength));
+  }
+  else if (geom->PlaneShape())
+  {
+    const auto plane = geom->PlaneShape();
+    const auto normal = convertVec(math::eigen3::convert(plane->Normal()));
+    shape = new btStaticPlaneShape(normal, 0);
+  }
+
+  // TODO(lobotuerk) find if there's a  way to use friction and related
+  // info in bullet dynamics
+
+  // const auto &surfaceElement = _collision.Element()->GetElement("surface");
+  //
+  // // Get friction
+  // const auto &odeFriction = surfaceElement->GetElement("friction")
+  //                             ->GetElement("ode");
+  // const auto mu = odeFriction->Get<btScalar>("mu");
+  //
+  // // Get restitution
+  // const auto restitution = surfaceElement->GetElement("bounce")
+  //                             ->Get<btScalar>("restitution_coefficient");
+  if (shape != nullptr)
+  {
+    const auto &linkInfo = this->links.at(_linkID);
+    const auto &modelID = linkInfo->model;
+    const auto &modelInfo = this->models.at(modelID);
+    const auto &link = linkInfo->link;
+    const auto &world = this->worlds.at(modelInfo->world)->world;
+    // btTransform transform;
+    delete link->getCollisionShape();
```

Comment:
Please leave a comment here stating why it's solved this way here. Maybe that can change in the future

Reply:
Updated on 30ae58a, hopefully, this does not spam the console